[Bug] Fixed primal weather remaining active during transition from double to single battle

tlelievre26 opened the pull request:
<!-- Make sure the title includes categorization (i.e. [Bug], [QoL], [Localization]) -->
<!-- Make sure that this PR is not overlapping with someone else's work -->
<!-- Please try to keep the PR self-contained (and small) -->

## What are the changes?
Fixes a bug where primal weather would not go away when the mon with the ability is forced back into it's ball during the transition from a double to a single battle.

## Why am I doing these changes?
Relevant issue: https://github.com/pagefaultgames/pokerogue/issues/3385

## What did change?
<!-- Explicitly state what are the changes introduced by the PR -->
<!-- You can make use of a comparison between what was the state before and after your PR changes -->
The `ReturnPhase` class that controls the forced switch going from a double to a single battle now calls the `applyPreSwitchOutAbAttrs` function for controlling how abilities behave during a switch.

### Screenshots/Videos
<!-- If your change is changing anything on the user experience, please provide visual proofs of it -->
<!-- Please take screenshots/videos before and after your changes, to show what is brought by this PR -->

## How to test the changes?
<!-- How can a reviewer test your changes once they check out on your branch? -->
<!-- Did you just make use of the `src/overrides.ts` file? -->
<!-- Did you introduce any automated tests? -->
<!-- Do the reviewer need to do something special in order to test your change? -->
Used the session file from the relevant issue until I got to a double battle and observed that Desolate Land correctly got deactivated.

Unsure if this change would justify any automated tests so if I should add some please let me know!

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

DayKev commented:
Duplicate of #2243 ?

tlelievre26 commented:
> Duplicate of #2243 ?

Seems like it, nobody had mentioned working on it in the Discord issue so I wasn't aware there was another fix for it.

DayKev commented:
I notice you also added `this.lastPokemon = this.getPokemon();` whereas the other one didn't, is this also a necessary change to fix something?

tlelievre26 commented:
> I notice you also added `this.lastPokemon = this.getPokemon();` whereas the other one didn't, is this also a necessary change to fix something?

The other PR called `applyPreSwitchoutAbAttributes` in the `onEnd` function, which already has a variable defined for this:
https://github.com/pagefaultgames/pokerogue/blob/e46491098d9edb17db2fdb741190f873178971f7/src/phases.ts#L1742

I added the call to the `switchAndSummon` function instead because that's where the attributes get applied in the `SwitchSummonPhase`, which meant I needed to add a variable to hold `this.getPokemon()`

DayKev commented on src/phases/return-phase.ts:
```diff
@@ -1,4 +1,5 @@
 import BattleScene from "#app/battle-scene.js";
+import { applyPreSwitchOutAbAttrs, PreSwitchOutAbAttr } from "#app/data/ability.js";
```
```suggestion
import { applyPreSwitchOutAbAttrs, PreSwitchOutAbAttr } from "#app/data/ability";
```

DayKev commented:
> Unsure if this change would justify any automated tests so if I should add some please let me know!

Probably, so I added the test lol.

torranx commented:
I'm thinking should the test be in `weather_strong_winds.ts` instead?

DayKev commented:
> I'm thinking should the test be in `weather_strong_winds.ts` instead?

Possibly. The primals' abilities should have some tests eventually, and this weather is directly linked to Mega-Ray and its ability though.

frutescens commented:
Hi! Could you move the tests to arena/weather_strong_winds.test.ts instead?
